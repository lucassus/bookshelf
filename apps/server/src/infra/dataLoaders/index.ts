export { buildAuthorsLoader } from "./authorsLoader";
export { buildBooksLoader } from "./booksLoader";
export { buildUsersLoader } from "./usersLoader";
