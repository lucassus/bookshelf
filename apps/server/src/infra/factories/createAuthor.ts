import faker from "faker";

import { createEntity } from "./createEntity";
import { Author } from "~/infra/database/entity";

const AUTHOR_PHOTOS = [
  "/authors/JK%20Rowling.jpg",
  "/authors/James%20S.%20A.%20Corey.jpg",
  "/authors/Andrzej%20Sapkowski.jpg",
  "/authors/JRR%20Tolkien.jpg"
];

export type CreateAuthorAttributes = Partial<Author>;

export function createAuthor(
  attributes: CreateAuthorAttributes = {}
): Promise<Author> {
  return createEntity(Author, {
    name: faker.name.findName(),
    bio: faker.lorem.sentence(),
    photoPath: faker.random.arrayElement(AUTHOR_PHOTOS),
    ...attributes
  });
}
