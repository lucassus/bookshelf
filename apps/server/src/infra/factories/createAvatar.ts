import faker from "faker";

import { createEntity } from "./createEntity";
import { Avatar } from "~/infra/database/entity";

const AVATAR_IMAGES = [
  "/users/w13.png",
  "/users/m10.png",
  "/users/w2.png",
  "/users/m25.png"
];

const AVATAR_COLORS = [
  "red",
  "green",
  "blue",
  "yellow",
  "magenta",
  "pink",
  "black"
];

export type CreateAvatarAttributes = Partial<Avatar>;

export function createAvatar(
  attributes: CreateAvatarAttributes = {}
): Promise<Avatar> {
  return createEntity(Avatar, {
    imagePath: faker.random.arrayElement(AVATAR_IMAGES),
    color: faker.random.arrayElement(AVATAR_COLORS),
    ...attributes
  });
}
