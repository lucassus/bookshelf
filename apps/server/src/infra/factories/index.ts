export { createAuthor } from "./createAuthor";
export { createAvatar } from "./createAvatar";
export { createBook } from "./createBook";
export { createBookCopy } from "./createBookCopy";
export { createUser } from "./createUser";
export { createReview } from "./createReview";
