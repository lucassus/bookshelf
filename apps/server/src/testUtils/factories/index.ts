export { createAuthor } from "./createAuthor";
export { createAvatar } from "./createAvatar";
export { createBook } from "./createBook";
export { createBookCopy } from "./createBookCopy";
export { createUser } from "./createUser";
