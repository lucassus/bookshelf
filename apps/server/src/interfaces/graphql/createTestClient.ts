import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import cookie from "cookie";
import express from "express";
import httpMocks from "node-mocks-http";

import { AUTH_COOKIE_NAME } from "../../infra/config";
import { User } from "../../infra/database/entity";
import { generateAuthToken } from "../../infra/support/authentication";
import { createContext } from "./context";
import { rootSchema } from "./rootSchema";

export function createTestClient({
  currentUser,
  res = httpMocks.createResponse()
}: {
  currentUser?: User;
  res?: express.Response;
} = {}): ApolloServerTestClient {
  const config = {
    schema: rootSchema,
    context: () => {
      const req = httpMocks.createRequest();

      if (currentUser) {
        req.headers = {
          cookie: cookie.serialize(
            AUTH_COOKIE_NAME,
            generateAuthToken(currentUser)
          )
        };
      }

      return createContext({ req, res });
    }
  };

  return createApolloTestClient(new ApolloServer(config));
}
