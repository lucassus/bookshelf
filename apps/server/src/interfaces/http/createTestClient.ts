import cookie from "cookie";
import express from "express";
import request from "supertest";

import { api as apiRoutes } from "./api/routes";
import { AUTH_COOKIE_NAME } from "~/infra/config";
import { User } from "~/infra/database/entity";
import { generateAuthToken } from "~/infra/support/authentication";

export function createTestClient({
  authToken,
  currentUser
}: { currentUser?: User; authToken?: string } = {}) {
  const app = express();

  app.use((req, res, next) => {
    const token = authToken || (currentUser && generateAuthToken(currentUser));

    if (token) {
      req.headers = {
        cookie: cookie.serialize(AUTH_COOKIE_NAME, token)
      };
    }

    next();
  });

  app.use("/api", apiRoutes);

  return request(app);
}
