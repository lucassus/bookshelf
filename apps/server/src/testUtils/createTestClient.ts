import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import express from "express";
import httpMocks from "node-mocks-http";

import { generateAuthToken } from "../common/authentication";
import { AUTH_COOKIE_NAME } from "../config";
import { User } from "../database/entity";
import { buildContext } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient({
  currentUser,
  res = httpMocks.createResponse()
}: {
  currentUser?: User;
  res?: express.Response;
} = {}): ApolloServerTestClient {
  const config = {
    schema: rootSchema,
    context: () => {
      const req = httpMocks.createRequest();

      if (currentUser) {
        req.cookies = {
          [AUTH_COOKIE_NAME]: generateAuthToken(currentUser)
        };
      }

      return buildContext({ req, res });
    }
  };

  return createApolloTestClient(new ApolloServer(config));
}
