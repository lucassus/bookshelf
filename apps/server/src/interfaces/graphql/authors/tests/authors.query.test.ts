import { gql } from "apollo-server-express";

import { createAuthor, createBook } from "../../../../infra/factories";
import { createTestClient } from "../../createTestClient";

test("authors query", async () => {
  // Given
  const firstAuthor = await createAuthor();
  await createBook({ author: firstAuthor });
  await createBook({ author: firstAuthor });
  await createBook({ author: firstAuthor });

  const secondAuthor = await createAuthor();
  await createBook({ author: secondAuthor });
  await createBook({ author: secondAuthor });

  // When
  const res = await createTestClient().query({
    query: gql`
      query {
        authors {
          id
          name
          books {
            id
            title
            description
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeNull();
  expect(res.data).toMatchSnapshot();
});
