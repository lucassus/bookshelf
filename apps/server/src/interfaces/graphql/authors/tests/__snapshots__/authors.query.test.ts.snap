// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authors query 1`] = `
{
  "authors": [
    {
      "books": [
        {
          "description": "Natus aliquam iste rem.",
          "id": "Mi1Cb29r",
          "title": "Dolorem Alias Qui Asperiores",
        },
        {
          "description": "Ipsa reiciendis nam.",
          "id": "MS1Cb29r",
          "title": "Fugit Facilis Mollitia",
        },
        {
          "description": "Iste aliquid odit aperiam dolore voluptatibus nemo.",
          "id": "My1Cb29r",
          "title": "Sit Dolore",
        },
      ],
      "id": "MS1BdXRob3I=",
      "name": "Miss Sadie Deckow-Reynolds",
    },
    {
      "books": [
        {
          "description": "Tenetur hic delectus.",
          "id": "NC1Cb29r",
          "title": "Dolores In Quae",
        },
        {
          "description": "Fuga adipisci reprehenderit fuga.",
          "id": "NS1Cb29r",
          "title": "Laboriosam Aliquam Aut Quasi",
        },
      ],
      "id": "Mi1BdXRob3I=",
      "name": "Catherine Mosciski",
    },
  ],
}
`;
